### PR TITLE
JSRT changes to support shim improvements

### PR DESCRIPTION
**Pull request by MikeHolman**
This change adds a laundry list of new (still very experimental) JSRT APIs, to support improvements in chakra shim functionality and performance.

### CODE REVIEW

File: lib/Jsrt/ChakraCommon.h [hunk 1]
```diff
@@ -2434,6 +2496,22 @@ typedef unsigned short uint16_t;
...
+    /// </summary>
+    /// <remarks>
+    ///     Requires an active script context.
+    /// </remarks>
+    /// <param name="object">The object to test.</param>
+    /// <param name="isConstructor">If the object is a constructor, <c>true</c>, <c>false</c> otherwise.</param>
+    /// <returns>
+    ///     The code <c>JsNoError</c> if the operation succeeded, a failure code otherwise.
+    /// </returns>
+    CHAKRA_API
```
Comment (liminzhu) [inline lib/Jsrt/ChakraCommon.h]: Can we get this and the other typedefs in ChakraCore.h? ChakraCommon is for shared APIs between Chakra and ChakraCore.

File: lib/Jsrt/ChakraCore.h [hunk 2]
```diff
@@ -1002,6 +1028,290 @@ CHAKRA_API
...
+///     A callback for when the object is traced. May be null.
+/// <param name="finalizeCallback">
+///     A callback for when the object is finalized. May be null.
+/// </param>
+/// <param name="prototype">Prototype object or nullptr.</param>
+/// <param name="object">The new object.</param>
+/// <returns>
+///     The code <c>JsNoError</c> if the operation succeeded, a failure code otherwise.
+/// </returns>
+CHAKRA_API
```
Comment (liminzhu) [inline lib/Jsrt/ChakraCore.h]: There are a bunch of APIs like this that do a combination of things. Maybe just publish one API with a bunch of _In_opt_ params instead?

File: lib/Jsrt/ChakraCommon.h [hunk 3]
```diff
@@ -721,6 +739,14 @@ typedef unsigned short uint16_t;
     /// </param>
     typedef void (CHAKRA_CALLBACK *JsFinalizeCallback)(_In_opt_ void *data);
 
+    /// <summary>
+    ///     A trace callback.
+    /// </summary>
+    /// <param name="data">
+    ///     The external data that was passed in when creating the object being traced.
+    /// </param>
+    typedef void (CHAKRA_CALLBACK *JsTraceCallback)(_In_opt_ void *data);
```
Comment (liminzhu) [inline lib/Jsrt/ChakraCommon.h]: Missed this guy - should be in chakracore.h

Comment (MikeHolman) [reply lib/Jsrt/ChakraCommon.h]: I tried moving it to ChakraCore.h, but we still use this typedef in ChakraFull, so it seemed kind of annoying. I guess I can #ifdef out all the places though

Comment (liminzhu) [reply lib/Jsrt/ChakraCommon.h]: hmmmm, i thought the callback is only used with JsCreateTracedExternalObject and JsCreateCustomExternalObject which is in chakracore.h? Btw is there any reason these two API are not merged into one?

Comment (MikeHolman) [reply lib/Jsrt/ChakraCommon.h]: I guess the APIs for these could be merged, but underneath they create very different types of objects so I'd rather hold off for now.

We share code for JsrtExternalObjects between ChakraCore and ChakraFull, and even though trace callback is only used in ChakraCore, it was referenced there in both builds... I moved it all behind ifdef now though

Comment (liminzhu) [reply lib/Jsrt/ChakraCommon.h]: SGTM, thx!

**Pull request merged**